util/kmod/kmod.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build linux
+// +build linux
+
+// Package kmod provides a simple API to attempt to ensure that a kernel
+// module is loaded in a wide variety of environments, and otherwise
+// report descriptive loggable error strings.
+// This package does not have extensive unit testing, as the broader set
+// of challenges associated with the package come from a wide variety of
+// distribution and linux version differences that are problematic to
+// mock/stub/emulate, including syscall boundary behaviors. The program
+// `ensuremod` is kept nearby the source that provides a method for
+// integration testing.
+package kmod
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+	"kernel.org/pub/linux/libs/security/libcap/cap"
+	"pault.ag/go/modprobe"
+)
+
+// hasKernelModule attempts to find a kernel module by name using procfs and
+// sysfs. If the module is found to be loaded, true is returned, in all other
+// cases false is returned, regardless of errors or a missing module.
+func hasKernelModule(name string) (bool, error) {
+	if _, err := os.Stat(filepath.Join("/sys/module", name)); err == nil {
+		return true, nil
+	}
+	procMod, err := os.Open("/proc/modules")
+	if err != nil {
+		return false, fmt.Errorf("module %s not found in /sys/module or /proc/modules: %w", name, err)
+	}
+	defer procMod.Close()
+
+	prefix := make([]byte, len(name)+1)
```

Comment:
`prefix := mem.S(name + " ")` is one alloc (for the string concat, not the `mem.S`)

and then below:

`if mem.HasPrefix(mem.B(line), prefix) { .... }`

Reply:
(that's using `go4.org/mem` which we use pretty often in the tree)

Reply:
Neat, nice package, thanks.